### PR TITLE
Fix Messages UI

### PR DESCRIPTION
**Pull request by roberthein**
New setup for the message cells where all complicated message cells use a "base cell" for the general layout and functionality.

### CODE REVIEW

File: Toshi/Controllers/Messaging/ChatController.swift [hunk 1]
```diff
@@ -437,6 +439,11 @@ final class ChatController: OverlayController {
         signalMessage.save()
 
         tableView.reloadRows(at: [indexPath], with: .fade)
+        
+        
+        
+        //todo maybe check this later?
```
Comment (yuliaveres) [inline Toshi/Controllers/Messaging/ChatController.swift]: check what ?

File: Toshi/Controllers/Messaging/ChatController.swift [hunk 2]
```diff
@@ -630,102 +625,97 @@ final class ChatController: OverlayController {
     }
 }
 
+extension ChatController: UITableViewDelegate {
+    
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        
+        if tableView.cellForRow(at: indexPath) is MessagesImageCell {
```
Comment (yuliaveres) [inline Toshi/Controllers/Messaging/ChatController.swift]: it is better to check model, datasource of messages, `.type == .image` than calling `cellForRow` explicitly

File: Toshi/Controllers/Messaging/ChatController.swift [hunk 3]
```diff
@@ -630,102 +625,97 @@ final class ChatController: OverlayController {
...
+        switch type {
+        case .simple:
+            return MessagesTextCell.reuseIdentifier
+        case .image:
+            return MessagesImageCell.reuseIdentifier
+        case .paymentRequest, .payment:
+            return MessagesPaymentCell.reuseIdentifier
+        case .status:
+            return MessagesTextCell.reuseIdentifier
         }
```
Comment (yuliaveres) [inline Toshi/Controllers/Messaging/ChatController.swift]: duplication -> same as `.simple`

File: Toshi/Controllers/Messaging/ChatController.swift [hunk 4]
```diff
@@ -630,102 +625,97 @@ final class ChatController: OverlayController {
...
         guard let indexPath = self.tableView.indexPath(for: cell) as IndexPath? else { return }
         guard let message = self.viewModel.messageModels.element(at: indexPath.row) as MessageModel? else { return }
-
+        
         adjustToPaymentState(.pendingConfirmation, at: indexPath)
-
+        
         guard let paymentRequest = message.sofaWrapper as? SofaPaymentRequest else { return }
-
+
```
Comment (yuliaveres) [inline Toshi/Controllers/Messaging/ChatController.swift]: whitespaces 😢

File: Toshi/Controllers/Messaging/ChatViewModel.swift [hunk 5]
```diff
@@ -77,14 +82,14 @@ final class ChatViewModel {
...
-    private(set) var messages: [Message] = [Message]() {
+    private(set) var messages: [Message] = [] {
         didSet {
             self.output.didReload()
         }
     }
     
     var visibleMessages: [Message] {
-        return self.messages.filter { (message) -> Bool in
+        return self.messages.filter { message -> Bool in
```
Comment (yuliaveres) [inline Toshi/Controllers/Messaging/ChatViewModel.swift]: ❤️

File: Toshi/Controllers/Messaging/ChatViewModel.swift [hunk 6]
```diff
@@ -62,13 +62,18 @@ final class ChatViewModel {
         return ChatsInteractor(output: self.output, thread: self.thread)
     }()
     
-    private(set) var contact: TokenUser? {
-        get {
-            return self.contactsManager?.tokenContact(forAddress: self.thread.contactIdentifier())
-        }
-        set {}
+    var contact: TokenUser? {
```
Comment (yuliaveres) [inline Toshi/Controllers/Messaging/ChatViewModel.swift]: at least `private(set)` ? 🤔

File: Toshi/MessagesUI/Cells/MessagesBasicCell.swift [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+                bubbleLeftConstraint?.isActive = false
+                bubbleRightConstantConstraint?.isActive = false
+                bubbleRightConstraint?.isActive = true
+                bubbleLeftConstantConstraint?.isActive = true
+            }
+        }
+    }
+    
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
```
Comment (yuliaveres) [inline Toshi/MessagesUI/Cells/MessagesBasicCell.swift]: why fatal error ?? seriously, just setup cell as in other init case ....... 😥

File: Toshi/MessagesUI/Cells/MessagesImageCell.swift [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+class MessagesImageCell: MessagesBasicCell {
+    
+    static let reuseIdentifier = "MessagesImageCell"
+    
+    var messageImage: UIImage? = nil {
+        didSet {
+            guard let messageImage = messageImage else {
+                heightConstraint?.isActive = false
+                heightConstraint = nil
```
Comment (yuliaveres) [inline Toshi/MessagesUI/Cells/MessagesImageCell.swift]: line before return Pliiiiiz 😉

File: Toshi/MessagesUI/Cells/MessagesImageCell.swift [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+                return
+            }
+            
+            messageImageView.image = messageImage
+            
+            let aspectRatio: CGFloat = messageImage.size.height / messageImage.size.width
+            
+            heightConstraint?.isActive = false
```
Comment (yuliaveres) [inline Toshi/MessagesUI/Cells/MessagesImageCell.swift]: in both cases not active ?

File: Toshi/MessagesUI/Cells/MessagesImageCell.swift [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+        let view = UIImageView()
+        view.contentMode = .scaleAspectFit
+        
+        return view
+    }()
+    
+    private var heightConstraint: NSLayoutConstraint?
+    
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
```
Comment (yuliaveres) [inline Toshi/MessagesUI/Cells/MessagesImageCell.swift]: pliz remove all fatal errors when cells used from IB

File: Toshi/MessagesUI/Cells/MessagesPaymentCell.swift [hunk 11]
```diff
@@ -0,0 +1,174 @@
+import Foundation
+import UIKit
+import TinyConstraints
+
+protocol MessagesPaymentCellSelectionDelegate {
```
Comment (yuliaveres) [inline Toshi/MessagesUI/Cells/MessagesPaymentCell.swift]: i think we may come up with better naming, f.e. just removing "selection" component ?

File: Toshi/MessagesUI/Cells/MessagesPaymentCell.swift [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+    func declinePayment(for cell: MessagesPaymentCell)
+}
+
+class MessagesPaymentCell: MessagesBasicCell {
+    
+    var selectionDelegate: MessagesPaymentCellSelectionDelegate?
+    
+    static let reuseIdentifier = "MessagesPaymentCell"
+    
+    lazy var titleLabel: UILabel = {
```
Comment (yuliaveres) [inline Toshi/MessagesUI/Cells/MessagesPaymentCell.swift]: every component should be restricted from outside 'set', so `private(set)` 🤷‍♂️

File: Toshi/MessagesUI/Cells/MessagesPaymentCell.swift [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+                statusLabel.text = paymentState.stateText
+                
+                textBottomConstraint?.isActive = false
+                buttonBottomConstraint?.isActive = false
+                statusBottomConstraint?.isActive = true
+            }
+        }
+    }
+    
+    var textBottomConstraint: NSLayoutConstraint?
```
Comment (yuliaveres) [inline Toshi/MessagesUI/Cells/MessagesPaymentCell.swift]: privatize all of those vara

File: Toshi/MessagesUI/Cells/MessagesPaymentCell.swift [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+    var statusBottomConstraint: NSLayoutConstraint?
+    var buttonBottomConstraint: NSLayoutConstraint?
+    
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
```
Comment (yuliaveres) [inline Toshi/MessagesUI/Cells/MessagesPaymentCell.swift]: fatal 😱

File: Toshi/MessagesUI/Cells/MessagesPaymentCell.swift [hunk 15]
```diff
@@ -0,0 +1,174 @@
...
+        
+        titleLabel.text = nil
+        subtitleLabel.text = nil
+        messageLabel.text = nil
+        statusLabel.text = nil
+        paymentState = .none
+    }
+    
+    func approvePayment(_ button: IconLabelButton) {
+        selectionDelegate?.approvePayment(for: self)
```
Comment (yuliaveres) [inline Toshi/MessagesUI/Cells/MessagesPaymentCell.swift]: we should use did component in naming.. `paymentCell: MessagePaymentCell, didPressApprove / didApprove`

File: Toshi/MessagesUI/Cells/MessagesStatusCell.swift [hunk 16]
```diff
@@ -0,0 +1,20 @@
+import Foundation
+import UIKit
+import TinyConstraints
+
+class MessagesStatusCell: MessagesBasicCell {
+    
+    static let reuseIdentifier = "MessagesStatusCell"
+    
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
```
Comment (yuliaveres) [inline Toshi/MessagesUI/Cells/MessagesStatusCell.swift]: 😩

File: Toshi/MessagesUI/Cells/MessagesStatusCell.swift [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+    }
+    
+    override init(style: UITableViewCellStyle, reuseIdentifier: String?) {
+        super.init(style: style, reuseIdentifier: reuseIdentifier)
+    }
+    
+    override func prepareForReuse() {
+        super.prepareForReuse()
```
Comment (yuliaveres) [inline Toshi/MessagesUI/Cells/MessagesStatusCell.swift]: remove this implementation please 😏

File: Toshi/MessagesUI/Cells/MessagesTextCell.swift [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+            bubbleView.backgroundColor = isOutGoing ? Theme.tintColor : Theme.incomingMessageBackgroundColor
+            
+            detectUsernameLinksIfNeeded()
+        }
+    }
+    
+    fileprivate let usernameDetector = try! NSRegularExpression(pattern: " ?(@[a-zA-Z][a-zA-Z0-9_]{2,59}) ?", options: [.caseInsensitive, .useUnicodeWordBoundaries])
+    
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
```
Comment (yuliaveres) [inline Toshi/MessagesUI/Cells/MessagesTextCell.swift]: 😩

File: Toshi/MessagesUI/Views/IconLabelButton.swift [hunk 19]
```diff
@@ -0,0 +1,72 @@
+import Foundation
+import UIKit
+import TinyConstraints
+
+class IconLabelButton: UIControl {
```
Comment (yuliaveres) [inline Toshi/MessagesUI/Views/IconLabelButton.swift]: IconLabel? 🤔  Button is enought, as widely known, buttons has both icon and text, but `BorderedButton` - is 👍  for this case ?

File: Toshi/MessagesUI/Views/IconLabelButton.swift [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+    }
+    
+    var font: UIFont? = nil {
+        didSet {
+            iconLabelView.titleLabel.font = font
+        }
+    }
+    
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
```
Comment (yuliaveres) [inline Toshi/MessagesUI/Views/IconLabelButton.swift]: 😩

File: Toshi/MessagesUI/Views/IconLabelView.swift [hunk 21]
```diff
@@ -0,0 +1,70 @@
+import Foundation
+import UIKit
+import TinyConstraints
+
+class IconLabelView: UIView {
```
Comment (yuliaveres) [inline Toshi/MessagesUI/Views/IconLabelView.swift]: do we need this class for the view which has icon + text ? is not that what UIButton does have ?

**Comment by yuliaveres** [comment]
Also, the only issue I have noticed so far is payment process: if you have time, please investigate this:
1 user
2 user,
1 user sends a payment request for 2 user
2 user approves request
1 user now sees not payment message but payment request for the amount s/he requested, with Approve/Decline options 🤷‍♂️

File: Toshi/Controllers/Messaging/ChatController.swift (hunk 2 above)
Comment (roberthein) [inline Toshi/Controllers/Messaging/ChatController.swift]: 👍

File: Toshi/Controllers/Messaging/ChatViewModel.swift (hunk 6 above)
Comment (roberthein) [inline Toshi/Controllers/Messaging/ChatViewModel.swift]: No. This is a get only property. 😉

File: Toshi/Controllers/Messaging/ChatViewModel.swift (hunk 5 above)
Comment (roberthein) [inline Toshi/Controllers/Messaging/ChatViewModel.swift]: 😊

File: Toshi/MessagesUI/Cells/MessagesImageCell.swift (hunk 8 above)
Comment (roberthein) [inline Toshi/MessagesUI/Cells/MessagesImageCell.swift]: 👍

File: Toshi/MessagesUI/Cells/MessagesImageCell.swift (hunk 9 above)
Comment (roberthein) [inline Toshi/MessagesUI/Cells/MessagesImageCell.swift]: Yes, I deactivate the old constraint (if it exists) and then set `heightConstraint` with a new active constraint.

File: Toshi/MessagesUI/Cells/MessagesPaymentCell.swift (hunk 11 above)
Comment (roberthein) [inline Toshi/MessagesUI/Cells/MessagesPaymentCell.swift]: Done!

**Comment by roberthein** [comment]
Thanks everybody! ✨

**Pull request merged**